### ISSUE TITLE
Fact for detected modules?

### ISSUE THREAD

**Issue by andyshinn** [debops/ansible-hwraid#5]
I was thinking of a way to install my own scripts for certain cards and thought it might be useful to have a fact to rely on for external tasks / roles where I could do something like:

`when: my_card in hwraid_detected_devices`

I was trying to get this to work with just an additional task and registering the results of the package installation. I'm sure there is something much more elegant:

```
- name: Install packages for recognized RAID devices
  apt:
    name: '{{ item.1 }}'
    state: 'latest'
    install_recommends: False
  with_subelements:
    - hwraid_device_database
    - packages
  register: hwraid_device_result
  when: ((hwraid_register_release is defined and hwraid_register_release) and
         item.0.module not in hwraid_blacklist and
         item.0.module in hwraid_register_modules.stdout_lines)

- name: Create list fact of recognized RAID devices
  set_fact:
    hwraid_recognized_devices: '{{ hwraid_device_result.results | selectattr("item", "defined") | map(attribute="item") | list }}'

- debug: var=hwraid_recognized_devices
```

Thoughts?

**Comment by drybjed** [comment]
Good idea. You don't need to parse the register .results in a separate
task, just do it in Ansible fact template itself, taking care of using
prior results.
14 cze 2016 19:04 "Andy Shinn" <notifications@github.com> napisał(a):

**Comment by andyshinn** [comment]
I'm not sure what you mean by that?

**Comment by drybjed** [comment]
First of all, the first time the role is run and you get the registered results of the task, you can parse them easily in the Jinja template which you will use to create the Ansible fact and store it in `/etc/ansible/facts.d/hwraid.fact`, as a JSON data. In there you can parse the output, find out what you want and store it.

But on subsequent runs that output will probably be different if even show up at all. So, let's say you have a list of packages installed by the role and stored in the Ansible facts. If you just run the role again, it will overwrite the fact file with either wrong or missing data, destroying what you saved in the process.

To mitigate that, you need to re-iterate the current facts in the template, and merge that information with registered results. That way, if you run the role again, already stored data willl be preserved and any additional data will be appended to the existing list.

There are lots of examples of this in other DebOps roles, if you want to see how it's done.
